src/lua/datetime.lua

```diff
@@ -0,0 +1,623 @@
+
+local SECS_PER_DAY     = 86400
+local NANOS_PER_SEC    = 1000000000
+
+-- c-dt/dt_config.h
+
+-- Unix, January 1, 1970, Thursday
+local DT_EPOCH_1970_OFFSET = 719163
+
+
+local datetime_t = ffi.typeof('struct datetime')
+
+local function is_datetime(o)
+    return ffi.istype(datetime_t, o)
+end
+
+local function check_date(o, message)
+    if not is_datetime(o) then
+        return error(("%s: expected datetime, but received %s"):
+                     format(message, o), 2)
+    end
+end
+
+local function check_table(o, message)
+    if type(o) ~= 'table' then
+        return error(("%s: expected table %s"):format(message, o), 2)
+    end
+end
+
+local function check_str(s, message)
+    if not type(s) == 'string' then
+        return error(("%s: expected string, but received %s"):
+                     format(message, s), 2)
+    end
+end
+
+-- range may be of a form of pair {begin, end} or
+-- tuple {begin, end, negative}
+-- negative is a special value (so far) used for days only
+local function check_range(v, range, txt)
+    local len = #range
+    assert(len == 2 or len == 3)
+
+    local left, right, neg = unpack(range)
+    if neg == v or (v >= left and v <= right) then
+        return
+    end
+
+    if neg == nil then
+        error(('value %d of %s is out of allowed range [%d, %d]'):
+              format(v, txt, left, right), 2)
+    else
+        error(('value %d of %s is out of allowed range [%d, %d..%d]'):
+              format(v, txt, neg, left, right), 2)
+    end
+end
+
+local function nyi(msg)
+    local text = 'Not yet implemented'
+    if msg ~= nil then
+        text = ("%s : '%s'"):format(text, msg)
+    end
+    error(text, 3)
+end
+
+-- offset if __seconds__ of 1970-01-01 from 0000-01-01
+local SECS_EPOCH_OFFSET = (DT_EPOCH_1970_OFFSET * SECS_PER_DAY)
+
+-- convert from epoch related time to Rata Die related
+local function local_rd(secs)
+    return math_floor((secs + SECS_EPOCH_OFFSET) / SECS_PER_DAY)
+end
+
+-- convert UTC seconds to local seconds, adjusting by timezone
+local function local_secs(obj)
+    return obj.epoch + obj.tzoffset * 60
+end
+
+local function utc_secs(epoch, tzoffset)
+    return epoch - tzoffset * 60
+end
+
+-- get epoch seconds, shift to the local timezone
+-- adjust from 1970-related to 0000-related time
+-- then return dt in those coordinates (number of days
+-- since Rata Die date)
+local function local_dt(obj)
+    return builtin.tnt_dt_from_rdn(local_rd(local_secs(obj)))
+end
+
+local function normalize_nsec(secs, nsec)
+    if nsec < 0 then
+        secs = secs - 1
+        nsec = nsec + NANOS_PER_SEC
+    elseif nsec >= NANOS_PER_SEC then
+        secs = secs + 1
+        nsec = nsec - NANOS_PER_SEC
+    end
+    return secs, nsec
+end
+
+local function datetime_cmp(lhs, rhs)
+    if not is_datetime(lhs) or not is_datetime(rhs) then
+        return nil
+    end
+    local sdiff = lhs.epoch - rhs.epoch
+    return sdiff ~= 0 and sdiff or (lhs.nsec - rhs.nsec)
```

Comment:
You don't need `()` here.

Reply:
But they are harmless (and for such Lua newbies as myself it simplify reading).